library/packages/src/lib/y2packager/product_upgrade.rb

```diff
@@ -0,0 +1,171 @@
+      #     changing a module to a base product.
+      #
+      #  4) As the last attempt try to find the installed base product in the
+      #     available base products. It is very likely that SLES will be upgraded
+      #     to SLES, SLED to SLED and so on.
+      #
+      #  If no candidate product is found then it returns nil. That should happen
+      #  only when using completely incompatible products.
+      #
+      # @return [Y2Packager::Product,nil] the new upgraded product
+      def new_base_product
+        available = Y2Packager::Product.available_base_products
+        return nil if available.empty?
+
+        # just one product?
+        product = find_by_count(available)
+        return product if product
+
+        # found by solver?
+        product = find_by_solver
+        return product if product
+
+        # found by hardcoded mapping?
+        product = find_by_mapping(available)
+        return product if product
+
+        # just 1:1 product upgrade?
+        find_by_name(available)
+      end
+
+    private
+
+      # check the count of the new base products
+      # @param available [Array<Y2Packager::Product>] the available base products
+      # @return [Y2Packager::Product,nil] the new upgraded product
+      def find_by_count(available)
+        return nil unless available.size == 1
+
+        # only one base product available, we can upgrade only to this product
+        log.info("Only one base product available: #{available.first}")
+        available.first
+      end
+
+      # We do not know which available product might upgrade the installed product
+      # if the installation medium contains several products.
+      # Temporarily turn on the update mode to let the solver select the product for upgrade,
+      # this will correctly handle possible product renames specified via Obsoletes/Provides.
+      # @return [Y2Packager::Product,nil] the new upgraded product
+      def find_by_solver
+        # store the current resolvable states
+        Yast::Pkg.SaveState
+
+        # run the solver in the upgrade mode
+        Yast::Pkg.PkgUpdateAll({})
+        log_products
+
+        product = Y2Packager::Product.selected_base
+        # save the solver test case for easier debugging if no product upgrade was found
+        Yast::Pkg.CreateSolverTestCase("/var/log/YaST2/solver-product-upgrade") unless product
+
+        # restore the original resolvable states
+        Yast::Pkg.RestoreState
+        log_products
+
+        log.info("Upgraded base product found by solver: #{product.inspect}")
+        product
+      end
+
+      # find the upgrade product from the fallback mapping
+      # @param available [Array<Y2Packager::Product>] the available base products
+      # @return [Y2Packager::Product,nil] the new upgraded product
+      def find_by_mapping(available)
+        installed = Y2Packager::Product.installed_products
+
+        # sort the keys by length, try more products first
```

Comment:
it is good that you document what code did, but I miss important info, WHY? why more products have to be done earlier?

Reply:
I have added a comment with the reason.